libraries/chain/backing_store/db_key_value_format.cpp

```diff
@@ -16,37 +38,225 @@ namespace eosio { namespace chain { namespace backing_store { namespace db_key_v
+      if (comp != 0) {
+         return false;
+      }
+      auto start_offset = full_key.data() + sec_prefix_size;
+      const b1::chain_kv::bytes composite_primary_key {start_offset, start_offset + sizeof(primary_key)};
+      auto composite_loc = composite_primary_key.cbegin();
+      EOS_ASSERT(b1::chain_kv::extract_key(composite_loc, composite_primary_key.cend(), primary_key), bad_composite_key_exception,
+                 "DB intrinsic key-value store invariant has changed, extract_key should only fail if the string size is"
+                 " less than the sizeof(uint64_t)");
+      return true;
+   }
+
+   key_type extract_key_type(const b1::chain_kv::bytes& composite_key) {
+      key_type kt = key_type::primary;
+      std::tie(std::ignore, std::ignore, std::ignore, kt) = detail::extract_from_composite_key(composite_key.cbegin(), composite_key.cend());
+      return kt;
    }
 
+   key_type extract_primary_to_sec_key_type(const b1::chain_kv::bytes& composite_key) {
+      b1::chain_kv::bytes::const_iterator composite_loc;
+      key_type kt = key_type::primary;
+      std::tie(std::ignore, std::ignore, composite_loc, kt) = detail::extract_from_composite_key(composite_key.cbegin(), composite_key.cend());
+      EOS_ASSERT(kt == key_type::primary_to_sec, bad_composite_key_exception,
+                 "DB intrinsic extract_primary_to_sec_key_type was passed a key that was not of type: ${type}", ("type", detail::to_string(kt)));
+      EOS_ASSERT(composite_loc != composite_key.cend(), bad_composite_key_exception,
+                 "DB intrinsic extract_primary_to_sec_key_type was passed a key that was only the type prefix");
+      const key_type sec_kt{*composite_loc++};
+      return sec_kt;
+   }
+
+   intermittent_decomposed_values get_prefix_thru_key_type(const b1::chain_kv::bytes& composite_key) {
+      return detail::extract_from_composite_key(composite_key.cbegin(), composite_key.cend());
+   }
+
+   b1::chain_kv::bytes extract_legacy_key(const eosio::session::shared_bytes& complete_key) {
+      const auto slice = extract_legacy_slice(complete_key);
+      b1::chain_kv::bytes ret {slice.data(), slice.data() + slice.size()};
+      return ret;
+   }
+
+   rocksdb::Slice extract_legacy_slice(const eosio::session::shared_bytes& full_key) {
+      // offset into the whole key for the old postfix (after db type and contract)
+      return {full_key.data() + detail::db_type_and_code_size,
+              full_key.size() - detail::db_type_and_code_size};
+   }
+
+   eosio::session::shared_bytes create_full_key(const b1::chain_kv::bytes& composite_key, name code) {
+      static const char db_type_prefix = detail::get_rocksdb_contract_db_prefix_as_char();
+      b1::chain_kv::bytes code_as_bytes;
+      b1::chain_kv::append_key(code_as_bytes, code.to_uint64_t());
+      auto ret = eosio::session::make_shared_bytes<std::string_view, 3>({std::string_view{&db_type_prefix, 1},
```

Comment:
I think I removed the `make_shared_bytes` function.  We can add it back though since it would be much easier than converting all these over.

Reply:
I first tried making it one of the constructors, but was having lots of compiler problems so I just went the easy route.  I'll circle back if this shows up as a performance hotspot.